File: python/phonebook/timer.py
"""Faithful port of include/timer.hpp."""
import time
from typing import Callable

Repeats = 5


def time_it(work: Callable[[], None]) -> float:
    t0 = time.perf_counter()
    work()
    t1 = time.perf_counter()
    return (t1 - t0) * 1000.0


def benchmark(work: Callable[[], None], repeats: int = 5) -> float:
    if repeats <= 0:
        return 0.0
    best = float("inf")
    for _ in range(repeats):
        t = time_it(work)
        if t < best:
            best = t
    return best


def print_task_duration(work: Callable[[], None]) -> None:
    print(f"\nTook: {time_it(work)}ms.")

File: go/phonebook/timer.go
package main

import (
	"fmt"
	"math"
	"time"
)

// Faithful port of include/timer.hpp.

// TimeIt runs work and returns elapsed milliseconds.
func TimeIt(work func()) float64 {
	t0 := time.Now()
	work()
	return float64(time.Since(t0).Nanoseconds()) / 1e6
}

// Benchmark returns the best of `repeats` runs.
func Benchmark(work func(), repeats ...int) float64 {
	r := 5
	if len(repeats) > 0 {
		r = repeats[0]
	}
	if r <= 0 {
		return 0.0
	}
	best := math.MaxFloat64
	for i := 0; i < r; i++ {
		t := TimeIt(work)
		if t < best {
			best = t
		}
	}
	return best
}

// PrintTaskDuration prints "\nTook: Xms." like C++ printTaskDuration.
func PrintTaskDuration(work func()) {
	fmt.Printf("\nTook: %vms.\n", TimeIt(work))
}
